### PR TITLE
Creates extension for local csv and shp import

### LINKED ISSUE #341: Importer shapefile (.shp)
Il serait intéressant de permettre à l'utilisateur l'import d'un shapefile local (de taille raisonnable) comme couche temporaire pour la session active (comparable à l'import csv existant).

Des questions qui se posent sont :
* Serait-il mieux de le développer comme extension ou dans le cœur du code ? (probablement en utilisant une lib comme https://www.npmjs.com/package/shapefile)
* Si le dév était fait dans le cœur, est-ce qu'on choisirait de l'intégrer dans l'import csv pour avoir un outil d'import fichier ? (En cas du shp sans ouvrir la modal si on arrive à transformer le string du .prj en proj4js)
* Sinon, niveau UI il faudrait décider si on passe par une couche temporaire qui s'affiche dans le menu (comme l'import csv) ou un bouton qui ouvre le composant.

Merci pour vos retours sur le sujet.

### PR DESCRIPTION
**Pull request by tkohr**
This PR intends to resolve #341 by:
- moving the existing code for the current csv import (csv.js) into a new "fileimport" extension
- adding the capability to import zipped shapefiles (along with their .dbf and .prj) via the fileimport extension

The user can access the file import interface via the layertree (as before) or a new button in the toolbar (custom component).

The extension includes the following dependencies (for the shp import):
- https://github.com/gildas-lormeau/zip.js
- https://github.com/mbostock/shapefile

### CODE REVIEW

**Comment by Gaetanbrl** [comment]
Merci @tkohr.
Je check la PR dès que j'ai un moment.

**Comment by Gaetanbrl** [comment]
Here is my test : 

1.  Create shapefile with Qgis
2. Create test.zip from files : 

image


3. Open csv.xml demo
4. Clic to "Glisser un csv...."
5. I have the "options" windo with weirds infos : 

image

The system try to geocod the file @tkohr ?

**Comment by tkohr** [comment]
>     3. I get this "options" window :

Thanks for testing @Gaetanbrl , my check here must be to restrictive, I will try to fix this: https://github.com/camptocamp/mviewer/blob/GSGEODPT43-11-import-shp/demo/addons/fileimport/fileimport.js#L161 What's the mimetype of your file?

> The system try to geocod the file @tkohr ?

Yes I had the impression that interpreting the .prj content works pretty well with proj4js (https://github.com/camptocamp/mviewer/blob/GSGEODPT43-11-import-shp/demo/addons/fileimport/fileimport.js#L327) and I think a .prj file is mandatory for a shp. Maybe I am wrong and we need a fallback option to allow choosing an SRS manually!?

**Comment by Gaetanbrl** [comment]
> I think a .prj file is mandatory for a shp.

Some people receive SHP without .prj, so in real life we need to provide this use case.

B ythe way, Qgis allow to read SHP without .prj.

**Comment by Gaetanbrl** [comment]
> the mimetype of your file

http://mimeapplication.net/x-zip-compressed

File: demo/addons/fileimport/fileimport.js [hunk 1]
```diff
@@ -309,6 +304,78 @@ const fileimport = (function () {
         $("#srs-select").append(srs.join(" "));
     }
 
+    var _unzip = function(file, oLayer) {
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js:307]: Is it possible to separate the code in 3 files ?

main.js, csv.js, shp.js

File: demo/fileimport.xml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    </baselayers>     
+    <proxy url=""/>
+    <olscompletion url="http://api-adresse.data.gouv.fr/search/" type="ban"/>
+    <searchparameters bbox="false" localities="true" features="false" />   
+    <extensions>
+        <extension type="component" id="fileimport" path="demo/addons"/>
+    </extensions>
+    <themes> 
+        <theme name="Import" collapsed="false" id="import" icon="fas fa-users">
+            <layer type="import" id="csv1" name="demo csv"  visible="true"
```
Comment (Gaetanbrl) [inline demo/fileimport.xml:26]: You could directly create file with fileimport extension. User just need to import file.

But it's a good idea to manage this layer as normal layer. So why not...

File: demo/addons/fileimport/fileimport.js [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+            }
+        } else {
+            var alertText = mviewer.lang
+            ? mviewer.lang[mviewer.lang.lang]("fileimport.alert.config")
+            : "Pas de couche d'import configurée"
+            mviewer.alert(alertText, "alert-warning");
+        }
+    }
+
+    var _wizard_modal =
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js:23]: This file could be create separately and load with promise or simple ajax request.

1. Create template/fileimport.html  (contain this HTML code)
2. Add param into config file to register this file location 
3. In init, call from config url  and load this code to create modal

More clean but maybe not to fast.

File: demo/addons/fileimport/fileimport.js [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+                    formData.append("columns", value);
+                });
+            } else {
+                formData.append("citycode", oLayer.geocodingcitycode);
+            }
+            $("#csv-status").attr("class", "wait");
+            $.ajax({
+                type: "POST",
+                processData: false,
+                url: "https://api-adresse.data.gouv.fr/search/csv/",
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js:399]: Need to appear in config file as options.

File: demo/addons/fileimport/fileimport.html [hunk 5]
```diff
@@ -0,0 +1,3 @@
+  <button href="#" id="fileimport-button" title="Importer fichier local" i18n="fileimport.button.title" class="btn btn-default btn-raised">
+   <span class="glyphicon glyphicon-import" aria-hidden="true"></span>
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.html:2]: Don't work in mobile device ?

File: demo/addons/fileimport/fileimport.js [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+                // Use DataTransfer interface to access the file(s)
+                for (var i = 0; i < ev.dataTransfer.files.length; i++) {
+                    console.log('... file[' + i + '].name = ' + ev.dataTransfer.files[i].name);
+                }
+            }
+            // Pass event to removeDragData for cleanup
+            this.removeDragData(ev)
+        },
+        removeDragData: function (ev) {
+            console.log('Removing drag data')
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js:556]: console.log shall be removed.

File: demo/addons/fileimport/fileimport.js [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+            }
+        },
+        loadLocalFile: _loadLocalFile,
+        geocodeit: _geocodeit,
+        dragOverHandler: function (ev) {
+            // Prevent default behavior (Prevent file from being opened)
+            ev.preventDefault();
+        },
+        dropHandler: function (ev) {
+            console.log('File(s) dropped');
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js:533]: console.log shall be removed.

File: demo/addons/fileimport/fileimport.js [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+            _resetForms();
+            //remove existing features. Source can be used many times with differnet files
+            var _src = mviewer.getLayers()[idlayer].layer.getSource().clear();
+            var oLayer = mviewer.getLayers()[idlayer];
+            var reader = new FileReader();
+            //Constraint : file must be encoded in UTF-8 and first line is named fields
+            reader.readAsText(file, "UTF-8");
+            reader.onload = function (evt) {
+                //Show wizard modal
+                console.log("getfile");
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js]: console.log shall be removed.

File: demo/addons/fileimport/fileimport.js [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+                    var alertText = mviewer.lang
+                        ? mviewer.lang[mviewer.lang.lang]("csv.alert.geocoding")
+                        : "Problème avec le géocodage"
+                    mviewer.alert(alertText + thrownError, "alert-warning");
+                    $("#csv-status").attr("class", "start");
+                }
+            });
+
+        } else {
+            console.log("Ce geocoder " + oLayer.geocoder + " n'est pas pris en compte");
```
Comment (Gaetanbrl) [inline demo/addons/fileimport/fileimport.js:415]: I18n missing.

**Comment by Gaetanbrl** [comment]
Need to be fix and test again before merge @tkohr .

But it's a good start ! ;)

File: demo/addons/fileimport/fileimport.html (hunk 5 above)
Comment (tkohr) [reply demo/addons/fileimport/fileimport.html]: Indeed import does not seem very useful on mobile, so I hided the button for now.

File: demo/addons/fileimport/fileimport.js (hunk 4 above)
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js:461]: I think it's already present in param `geocoder` which includes more than the `url` (handling the response)

File: demo/addons/fileimport/fileimport.html (hunk 5 above)
Comment (Gaetanbrl) [reply demo/addons/fileimport/fileimport.html]: I agree.

File: demo/addons/fileimport/fileimport.js [hunk 10]
```diff
@@ -309,37 +309,40 @@ const fileimport = (function () {
...
-                        });
-                    });
+                    var shpfile = entries.filter(entry => entry.filename.split(".")[1] === "shp")[0];
+                    var dbffile = entries.filter(entry => entry.filename.split(".")[1] === "dbf")[0];
+                    var prjfile = entries.filter(entry => entry.filename.split(".")[1] === "prj")[0];
+                    var shpPromise = getFileContents(shpfile, "application/octet-stream");
+                    var dbfPromise = getFileContents(dbffile, "application/x-dbase");
+                    var prjPromise = getFileContents(prjfile, "text/txt");
+                    Promise.all([shpPromise, dbfPromise, prjPromise])
+                        .then(async ([shpResponse, dbfResponse, prjResponse]) => {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js]: Oh yeah I forgot, added a rejection for error handling

File: demo/addons/fileimport/fileimport.js [hunk 11]
```diff
@@ -309,37 +309,40 @@ const fileimport = (function () {
...
+                            var shpContents = await shpResponse.arrayBuffer();
+                            var dbfContents = await dbfResponse.arrayBuffer();
+                            var prjContents = await prjResponse.text();
+                            // register named projection for use in _loadShp
+                            var projId = `Proj-${Date.now()}`;
+                            proj4.defs(projId, prjContents);
```
[the remark this answers is not in the document]
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js:360]: Yes good point, I added a modal for the first use case

File: demo/addons/fileimport/fileimport.js [hunk 12]
```diff
@@ -155,106 +155,101 @@ const fileimport = (function () {
...
-            reader.onload = function (evt) {
-                //Show wizard modal
-                console.log("getfile");
-                $("#geocoding-modal").modal("show");
-                $("#geocoding-modal button.geocode").attr("data-layerid", idlayer);
-                //update layer title with file name
-                $("#geocoding-modal .csv-name").val(file.name);
-                //Parse csv file to convert it as json object
-                var tmp = Papa.parse(evt.target.result, { header: true });
+            if (file.type === "application/zip") {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js]: Thanks for bringing up this, I think the mimetype list should be complete now

File: demo/addons/fileimport/fileimport.js (hunk 4 above)
Comment (Gaetanbrl) [reply demo/addons/fileimport/fileimport.js:461]: So you prefer to keep hard URL code  ?

Just imagine base API URL change. Admin user (no dev profile) couldn't fix that.

I prefer to centralize all URL in config.

File: demo/addons/fileimport/fileimport.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply demo/addons/fileimport/fileimport.js]: For sure now ^^

File: demo/addons/fileimport/fileimport.js (hunk 1 above)
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js:342]: Yes good point, would be nice to have but I don't have the time today and will be less available the rest of the month due to holidays :-(

File: demo/fileimport.xml (hunk 2 above)
Comment (tkohr) [reply demo/fileimport.xml:26]: Yes I thought about it, but did not want to get rid of the existing option of adding a csv without the modal. Not sure, however if this option is very useful at all.

File: demo/addons/fileimport/fileimport.js (hunk 1 above)
Comment (Gaetanbrl) [reply demo/addons/fileimport/fileimport.js:342]: Ok ;)

File: demo/addons/fileimport/fileimport.js (hunk 3 above)
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js:25]: Sorry, but I also put it into the "nice to have" refacto

Comment (Gaetanbrl) [reply demo/addons/fileimport/fileimport.js:25]: > "nice to have"
I agree.

**Comment by tkohr** [comment]
Thanks a lot for the review and testing @Gaetanbrl . I tried to fix the blocking issues. Feel free to test / comment

**Comment by Gaetanbrl** [comment]
> Thanks a lot for the review and testing @Gaetanbrl . I tried to fix the blocking issues. Feel free to test / comment

Yep, i will test and try soon, unfortunately not today...

File: demo/addons/fileimport/fileimport.js (hunk 4 above)
Comment (Gaetanbrl) [reply demo/addons/fileimport/fileimport.js:461]: > Taken the fact that the import only supports one geocoder

It's true.

> we could perhaps use the geocoder param

Is it possible to use geocoder without xml config ? If not, use geocoder param. Else use addon config.

But you could by default use addon URL and read config.xml param if set.... just an idea.

File: demo/addons/fileimport/README.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
+## Import de fichier
+
+Cette extension permet d'importer des fichiers local. Les formats supportés actuellement sont `csv` et `shp`. 
+Dans le cas du Shapefile, le fichier `.shp` doit se trouver compressé dans un fichier `.zip` qui inclut également 
+un fichier `.dbf` (pour les attributs) et `.prj` (pour permettre l'interprétation du SRS à l'extension).
```
Comment (RemiDesgrange) [inline demo/addons/fileimport/README.md:5]: Peut-être ajouter que les fichiers sont chargé localement dans le browser de l'utilisateur, et que cette visualisation n'est pas persisté sur le serveur (ou dans le browser) ?

File: demo/addons/fileimport/featurestyles.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      })
+      }),
+      'Polygon': new ol.style.Style({
+      fill: new ol.style.Fill({
+          color: _importFillColor
+      }),
+      stroke: new ol.style.Stroke({
+          color: _importStrokeColor,
+          width: _width
+      })
```
Comment (RemiDesgrange) [inline demo/addons/fileimport/featurestyles.js:34]: Petit soucis d'indentation ?

File: demo/addons/fileimport/README.md (hunk 13 above)
Comment (Gaetanbrl) [reply demo/addons/fileimport/README.md:5]: > Cette extension permet d'importer des fichiers local

Je propose :

Cette extension permet d'importer des fichiers de façon temporaire (en local) des fichiers au format `csv` ou `shp`. Les données importées ne sont pas sauvegardées et seront perdues à la fermeture du navigateur.

File: demo/addons/fileimport/featurestyles.js (hunk 14 above)
Comment (tkohr) [reply demo/addons/fileimport/featurestyles.js]: Bien vu, merci

File: demo/addons/fileimport/fileimport.js (hunk 4 above)
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js]: Having had a look at it again, I added a param `geocoderurl' to allow an admin to override the url incase the ban url changes or he wants to use an API that behaves like the ban API. I kept it optional to minimize the mandatory params. In parallel I conserved the `geocoder` param to allow the implementation of a different API "type" in the future...

**Comment by tkohr** [comment]
@Gaetanbrl I had a "last" look at the PR, rebasing and treating the last open point (last two commits). Nothing more to add on my side...

**Comment by Gaetanbrl** [comment]
> Nothing more to add on my side...

ok i will check soon.

**Comment by tkohr** [comment]
Thanks again in advance for rechecking!

**Comment by Gaetanbrl** [comment]
As soon as possible @tkohr - Thanks.

**Review by Gaetanbrl — CHANGES_REQUESTED** [review]
Needs some minors modifications before merge.

However, easy to read, and great plugin folders structure.
Works fine ! thanks @tkohr :)

**Comment by Gaetanbrl** [comment]
@spelhate on peut attendre le merge de cette PR pour la release à venir ?

**Comment by tkohr** [comment]
Thanks @Gaetanbrl I will fix this tomorrow

**Comment by spelhate** [comment]
> @spelhate on peut attendre le merge de cette PR pour la release à venir ?

Ok.

File: demo/addons/fileimport/README.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
+## Import de fichier
+
+Cette extension permet d'importer des fichiers de façon temporaire (en local) des fichiers au format `csv` ou `shp`. 
+Les données importées ne sont pas sauvegardées et seront perdues à la fermeture du navigateur.
+Dans le cas du Shapefile, le fichier `.shp` doit se trouver compressé dans un fichier `.zip` qui inclut également 
+un fichier `.dbf` (pour les attributs) et `.prj` (pour permettre l'interprétation du SRS à l'extension).
+En cas d'absence du fichier `.prj` l'utilisateur est sollicité d'indiquer le SRS.
+
```
Comment (spelhate) [inline demo/addons/fileimport/README.md:8]: Préciser également que le fichier .dbf doit être encodé en UTF-8. Sinon, on a des caractères spéciaux mal décodés dans la fiche d'information.

File: demo/addons/fileimport/fileimport.js [hunk 16]
```diff
@@ -0,0 +1,633 @@
...
+    };
+
+    // Load local file
+    var _loadLocalFile = function (idlayer, file) {
+        if (!file) {
+            file = document.getElementById("loadcsv-" + idlayer).files[0];
+        }
+        if (file) {
+            //remove existing features. Source can be used many times with differnet files
+            var _src = mviewer.getLayers()[idlayer].layer.getSource().clear();
```
[the remark this answers is not in the document]
Comment (spelhate) [reply demo/addons/fileimport/fileimport.js:183]: just remove "`var _src = `"

**Review by spelhate — CHANGES_REQUESTED** [review]

File: demo/addons/fileimport/fileimport.js [hunk 17]
```diff
@@ -0,0 +1,633 @@
...
+                    var dbfPromise = getFileContents(dbffile, "application/x-dbase");
+                    if (prjfile) {
+                        //get projection from .prj file
+                        getFileContents(prjfile, "text/txt").then(async (prjResponse) => {
+                            var prjContents = await prjResponse.text();
+                            // register named projection for use in _loadShp
+                            var projId = `Proj-${Date.now()}`;
+                            proj4.defs(projId, prjContents);
+                            ol.proj.proj4.register(proj4);
+                            _resolveShpDbf(shpPromise, dbfPromise, projId, oLayer, shpfile.filename);
```
[the remark this answers is not in the document]
Comment (tkohr) [reply demo/addons/fileimport/fileimport.js:362]: Oh yeah, thanks for noticing

File: demo/addons/fileimport/README.md (hunk 15 above)
Comment (tkohr) [reply demo/addons/fileimport/README.md:8]: Ah oui bien vu

**Comment by tkohr** [comment]
Thanks again for all your reviewing @Gaetanbrl @spelhate . I added to commits to address your comments

**Review by Gaetanbrl — APPROVED** [review]
LGTM thanks @tkohr  ;)

**Comment by Gaetanbrl** [comment]
Wait @spelhate lasts return and we are good ! 👍

**Review by spelhate — APPROVED** [review]

**Comment by spelhate** [comment]
Thank's @tkohr & @Gaetanbrl for review.

**Pull request merged**